---
title: "TMRC2 `r Sys.getenv('VERSION')`: Promastigote (mostly) Differential Expression Analyses."
author: "atb abelew@gmail.com"
date: "`r Sys.Date()`"
output:
  html_document:
    code_download: true
    code_folding: show
    fig_caption: true
    fig_height: 7
    fig_width: 7
    highlight: zenburn
    keep_md: false
    mode: selfcontained
    number_sections: true
    self_contained: true
    theme: readable
    toc: true
    toc_float:
      collapsed: false
      smooth_scroll: false
---

<style type="text/css">
body .main-container {
  max-width: 1600px;
}
body, td {
  font-size: 16px;
}
code.r{
  font-size: 16px;
}
pre {
  font-size: 16px
}
</style>

```{r options, include = FALSE}
library(Heatplus)
library(hpgltools)
library(glue)
tt <- try(devtools::load_all("~/hpgltools"))
knitr::opts_knit$set(progress = TRUE, verbose = TRUE, width = 90, echo = TRUE)
knitr::opts_chunk$set(
  error = TRUE, fig.width = 8, fig.height = 8, fig.retina = 2,
  out.width = "100%", dev = "png",
  dev.args = list(png = list(type = "cairo-png")))
old_options <- options(digits = 4, stringsAsFactors = FALSE, knitr.duplicate.label = "allow")
ggplot2::theme_set(ggplot2::theme_bw(base_size = 12))
ver <- Sys.getenv("VERSION")
previous_file <- ""
rundate <- format(Sys.Date(), format = "%Y%m%d")

rmd_file <- "03differential_expression.Rmd"
savefile <- gsub(pattern = "\\.Rmd", replace = "\\.rda\\.xz", x = rmd_file)
loaded <- load(file = glue("rda/tmrc2_data_structures-v{ver}.rda"))
excel_out <- "analyses/transcriptome"
```

# Changelog

202405: Changed excel output directory to match organization scheme in box.
        Generally this means files go to
        analyses/transcriptome/{type_of_contrast}/{date}/something_{suffix}.xlsx
        Where suffix is _table for the full tables and _sig for the significant genes and
        will include information about whether sva etc was used.
202405: Adding some goseq results.

# Contrasts

** Note!  ** The new definitions of susceptible/resistant are tighter
than ever before, as a result there are no longer any ambiguous
samples.  Thus I removed the ambiguous contrasts in the following block.

```{r}
zymodeme_keeper <- list(
  "zymodeme" = c("z23", "z22"))
susceptibility_keepers <- list(
  "resistant_sensitive" = c("resistant", "sensitive"))
##    "resistant_ambiguous" = c("resistant", "ambiguous"),
##    "sensitive_ambiguous" = c("sensitive", "ambiguous"))
```

# Parasite Ontology data

Just a reminder that in data_structures.Rmd I created lp_go and lp_lengths

## Zymodeme enzyme gene IDs

Najib read me an email listing off the gene names associated with the zymodeme
classification.  I took those names and cross referenced them against the
Leishmania panamensis gene annotations and found the following:

They are:

1. ALAT: LPAL13_120010900 -- alanine aminotransferase
2. ASAT: LPAL13_340013000 -- aspartate aminotransferase
3. G6PD: LPAL13_000054100 -- glucase-6-phosphate 1-dehydrogenase
4. NH: LPAL13_14006100, LPAL13_180018500 -- inosine-guanine nucleoside hydrolase
5. MPI: LPAL13_320022300 (maybe) -- mannose phosphate isomerase (I chose phosphomannose isomerase)

Given these 6 gene IDs (NH has two gene IDs associated with it), I can do some
looking for specific differences among the various samples.

### Expression levels of zymodeme genes

The following creates a colorspace (red to green) heatmap showing the observed
expression of these genes in every sample.

```{r}
my_genes <- c("LPAL13_120010900", "LPAL13_340013000", "LPAL13_000054100",
              "LPAL13_140006100", "LPAL13_180018500", "LPAL13_320022300",
              "other")
my_names <- c("ALAT", "ASAT", "G6PD", "NHv1", "NHv2", "MPI", "other")

zymo_six_genes <- exclude_genes_expt(lp_two_strains, ids = my_genes, method = "keep")
strain_norm <- normalize_expt(zymo_six_genes, convert = "rpkm", filter = TRUE, transform = "log2")

zymo_heatmap <- plot_sample_heatmap(strain_norm, row_label = my_names)
zymo_heatmap

lp_norm <- normalize_expt(lp_two_strains, filter = TRUE, convert = "rpkm",
                          norm = "quant", transform = "log2")
zymo_heatmap_all <- plot_sample_heatmap(lp_norm)
zymo_heatmap_all
```

## Compare to highly expressed, variant genes

I want to compare the above heatmap with one which is comprised of all
genes with some 'significantly high' expression value and also a
not-negligible coefficient of variance.

```{r}
zymo_high_genes <- normalize_expt(lp_two_strains, filter = "cv", cv_min = 0.9)

high_strain_norm <- normalize_expt(zymo_high_genes, convert = "rpkm",
                                   norm = "quant", transform = "log2")
zymo_heatmap <- plot_sample_heatmap(high_strain_norm, row_label = my_names)
zymo_heatmap
```

I think this plot suggests that the difference between the two primary
strains is not really one of a few specific genes, but instead a
global pattern.

# Zymodeme differential expression

## No attempt at batch estimation

```{r}
zymo_de_nobatch <- all_pairwise(lp_zymo, filter = TRUE, model_batch = FALSE)
zymo_de_nobatch
## Including the plots causes the rda file to balloon to 3.4Gb in the following invocation.
## Removing them results in... holy crap 2.1Mb
zymo_table_nobatch <- combine_de_tables(
    zymo_de_nobatch, keepers = zymodeme_keeper,
    rda = glue("rda/zymo_tables_nobatch-v{ver}.rda"),
    excel = glue("{excel_out}/DE_Strain/{ver}/zymo_tables_nobatch-v{ver}.xlsx"))
zymo_table_nobatch
zymo_sig_nobatch <- extract_significant_genes(
    zymo_table_nobatch,
    according_to = "deseq", current_id = "GID", required_id = "GID",
    gmt = glue("excel/zymodeme_nobatch-v{ver}.gmt"),
    excel = glue("{excel_out}/DE_Strain/{ver}/zymo_sig_nobatch_deseq-v{ver}.xlsx"))
zymo_sig_nobatch
```

#### Gene ontology comparing the strains

There are too few genes at our current stringencies for a meaningful result.

```{r}
increased_z22 <- zymo_sig_nobatch[["deseq"]][["downs"]][["zymodeme"]]
increased_z23 <- zymo_sig_nobatch[["deseq"]][["ups"]][["zymodeme"]]
z22_goseq <- simple_goseq(increased_z22, go_db = lp_go, length_db = lp_lengths)
z23_goseq <- simple_goseq(increased_z23, go_db = lp_go, length_db = lp_lengths)
```

### Plot DE genes without batch estimation/adjustment

```{r}
zymo_table_nobatch[["plots"]][["zymodeme"]][["deseq_ma_plots"]]
zymo_table_nobatch[["plots"]][["zymodeme"]][["deseq_vol_plots"]]
```

Log ratio, mean average plot and volcano plot of the comparison of the
two primary zymodeme transcriptomes.  When the transcriptomes of the
two main strains (43 and 41 samples of z2.3 and z2.1) were compared
without any attempt at batch/surrogate estimation with DESeq2, 45 and
85 genes were observed as significantly higher in strain z2.3 and z2.2
respectively using a cutoff of 1.0 logFC and 0.05 FDR adjusted
p-value.  There remain a large number of genes which are likely
significantly different between the two strains, but fall below the
2-fold difference required for 'significance.'  This follows prior
observations that the parasite transcriptomes are constituitively
expressed.

When the same data was plotted via a volcano plot, the relatively
small range of fold changes compared to the large range of adjusted
p-values is visible.

## Attempt SVA estimate

```{r}
zymo_de_sva <- all_pairwise(lp_zymo, filter = TRUE, model_batch = "svaseq")
zymo_de_sva
zymo_table_sva <- combine_de_tables(
    zymo_de_sva, keepers = zymodeme_keeper,
    rda = glue("rda/zymo_tables_sva-v{ver}.rda"),
    excel = glue("{excel_out}/DE_Strain/{ver}/zymo_tables_sva-v{ver}.xlsx"))
zymo_table_sva
zymo_sig_sva <- extract_significant_genes(
    zymo_table_sva,
    according_to = "deseq",
    current_id = "GID", required_id = "GID",
    gmt = glue("excel/zymodeme_sva-v{ver}.gmt"),
    excel = glue("{excel_out}/DE_Strain/{ver}/zymo_sig_sva-v{ver}.xlsx"))
zymo_sig_sva
```

#### Gene ontology comparing the strains

There are too few genes at our current stringencies for a meaningful result.

```{r}
increased_z22 <- zymo_sig_sva[["deseq"]][["downs"]][["zymodeme"]]
increased_z23 <- zymo_sig_sva[["deseq"]][["ups"]][["zymodeme"]]
z22_goseq <- simple_goseq(increased_z22, go_db = lp_go, length_db = lp_lengths)
z23_goseq <- simple_goseq(increased_z23, go_db = lp_go, length_db = lp_lengths)
```

### Plot zymodeme DE genes with sva batch estimation/adjustment

When estimates from SVA were included in the statistical model used by
EdgeR, DESeq2, and limma; a nearly identical view of the data emerged.
I think this shows with a high degree of confidence, that sva is not
having a significant effect on this dataset.

```{r}
zymo_table_sva[["plots"]][["zymodeme"]][["deseq_ma_plots"]]
zymo_table_sva[["plots"]][["zymodeme"]][["deseq_vol_plots"]]
```

# Parasite Susceptibility to Drug (Current)

This susceptibility comparison is using the 'current' dataset.

Note again: we no longer have any ambiguous samples, so I commented
out a portion of the following block.

```{r}
sus_de_nobatch <- all_pairwise(lp_susceptibility, filter = TRUE, model_batch = FALSE)
sus_de_nobatch
sus_table_nobatch <- combine_de_tables(
  sus_de_nobatch, keepers = susceptibility_keepers,
  rda = glue("rda/sus_tables_nobatch-v{ver}.rda"),
  excel = glue("{excel_out}/DE_Susceptibility/{ver}/sus_tables_nobatch-v{ver}.xlsx"))
sus_table_nobatch
sus_sig_nobatch <- extract_significant_genes(
  sus_table_nobatch,
  excel = glue("{excel_out}/DE_Susceptibility/{ver}/sus_sig_nobatch-v{ver}.xlsx"))

sus_de_sva <- all_pairwise(lp_susceptibility, filter = TRUE, model_batch = "svaseq")
sus_de_sva
sus_table_sva <- combine_de_tables(
    sus_de_sva, keepers = susceptibility_keepers,
    rda = glue("rda/sus_tables_sva-v{ver}.rda"),
    excel = glue("{excel_out}/DE_Susceptibility/{ver}/sus_tables_sva-v{ver}.xlsx"))
sus_table_sva
sus_sig_sva <- extract_significant_genes(
  sus_table_sva, according_to = "deseq",
    excel = glue("{excel_out}/DE_Susceptibility/{ver}/sus_sig_sva-v{ver}.xlsx"))
sus_sig_sva

## To get a more true sense of sensitive vs resistant with sva, we kind of need to get rid of the
## unknown samples and perhaps the ambiguous.
## no_ambiguous <- subset_expt(lp_susceptibility, subset = "condition!='ambiguous'") %>%
##   subset_expt(subset = "condition!='unknown'")
## no_ambiguous_de_sva <- all_pairwise(no_ambiguous, filter = TRUE, model_batch = "svaseq")
## no_ambiguous_de_sva
## Let us see if my keeper code will fail hard or soft with extra contrasts...
## no_ambiguous_table_sva <- combine_de_tables(
##     no_ambiguous_de_sva, keepers = susceptibility_keepers,
##     excel = glue("excel/no_ambiguous_tables_sva-v{ver}.xlsx"))
## no_ambiguous_table_sva
## no_ambiguous_sig_sva <- extract_significant_genes(
##     no_ambiguous_table_sva, according_to = "deseq",
##     excel = glue("excel/no_ambiguous_sig_sva-v{ver}.xlsx"))
## no_ambiguous_sig_sva
```

#### Gene ontology comparing the susceptibility

```{r}
sus_sig_sva
increased_resistant <- sus_sig_sva[["deseq"]][["ups"]][["resistant_sensitive"]]
increased_sensitive <- sus_sig_sva[["deseq"]][["downs"]][["resistant_sensitive"]]
resistant_goseq <- simple_goseq(increased_resistant, go_db = lp_go, length_db = lp_lengths)
sensitive_goseq <- simple_goseq(increased_sensitive, go_db = lp_go, length_db = lp_lengths)
```

### Plot Susceptibility DE genes with sva batch estimation/adjustment

```{r}
sus_table_nobatch[["plots"]][["resistant_sensitive"]][["deseq_ma_plots"]]
sus_table_nobatch[["plots"]][["resistant_sensitive"]][["deseq_vol_plots"]]

sus_table_sva[["plots"]][["resistant_sensitive"]][["deseq_ma_plots"]]
sus_table_sva[["plots"]][["resistant_sensitive"]][["deseq_vol_plots"]]
```

Given that resistance/sensitivity tends to be correlated with strain,
one might expect similar results.  One caveat in this context though:
there are fewer strains with resistance/sensitivity definitions.  This
when the analysis was repeated without the ambiguous/unknown samples,
a few more genes were observed as significant.

# Comparing DE results from strain/sensitivity

```{r}
## zymo_table_sva[["plots"]][["zymodeme"]][["deseq_ma_plots"]][["plot"]]
zy_df <- zymo_table_sva[["data"]][["zymodeme"]]
sus_df <- sus_table_sva[["data"]][["resistant_sensitive"]]

both_df <- merge(zy_df, sus_df, by = "row.names")
plot_df <- both_df[, c("deseq_logfc.x", "deseq_logfc.y")]
rownames(plot_df) <- both_df[["Row.names"]]
colnames(plot_df) <- c("z23_vs_z22", "sensitive_vs_resistant")

compare <- plot_linear_scatter(plot_df)
pp(file = "images/compare_sus_zy.png")
compare$scatter
dev.off()
compare$scatter
compare$cor
```

# Parasite Susceptibility to Drug (Historical)

This susceptibility comparison is using the historical dataset.

```{r}
sushist_de_nobatch <- all_pairwise(lp_susceptibility_historical, filter = TRUE,
                                   model_batch = FALSE)
sushist_de_nobatch
sushist_table_nobatch <- combine_de_tables(
  sushist_de_nobatch, keepers = susceptibility_keepers,
  excel = glue("{excel_out}/DE_Susceptibility/sushist_tables_nobatch-v{ver}.xlsx"))
sushist_table_nobatch
sushist_sig_nobatch <- extract_significant_genes(
  sushist_table_nobatch,
  excel = glue("{excel_out}/DE_Susceptibility/sushist_sig_nobatch-v{ver}.xlsx"))
sushist_sig_nobatch

sushist_de_sva <- all_pairwise(lp_susceptibility_historical, filter = TRUE,
                               model_batch = "svaseq")
sushist_de_sva
sushist_table_sva <- combine_de_tables(
  sushist_de_sva, keepers = susceptibility_keepers,
  excel = glue("{excel_out}/DE_Susceptibility/sushist_tables_sva-v{ver}.xlsx"))
sushist_table_sva
sushist_sig_sva <- extract_significant_genes(
  sushist_table_sva, according_to = "deseq",
  excel = glue("{excel_out}/DE_Susceptibility/sushist_sig_sva-v{ver}.xlsx"))
sushist_sig_sva
```

# Cure/Fail association

```{r}
##cf_nb_input <- subset_expt(cf_expt, subset="condition!='unknown'")
cf_de_nobatch <- all_pairwise(lp_cf_known, filter = TRUE, model_batch = FALSE)
cf_de_nobatch
cf_table_nobatch <- combine_de_tables(
  cf_de_nobatch,
  excel = glue("{excel_out}/DE_Cure_vs_Fail/{ver}/cf_tables_nobatch-v{ver}.xlsx"))
cf_table_nobatch
cf_sig_nobatch <- extract_significant_genes(
  cf_table_nobatch,
  excel = glue("{excel_out}/DE_Cure_vs_Fail/{ver}/cf_sig_nobatch-v{ver}.xlsx"))
cf_sig_nobatch

cf_de <- all_pairwise(lp_cf_known, filter = TRUE, model_batch = "svaseq")
cf_de
cf_table <- combine_de_tables(
  cf_de,
  excel = glue("{excel_out}/DE_Cure_vs_Fail/{ver}/cf_tables-v{ver}.xlsx"))
cf_table
cf_sig <- extract_significant_genes(
  cf_table,
  excel = glue("{excel_out}/DE_Cure_vs_Fail/{ver}/cf_sig-v{ver}.xlsx"))
cf_sig
```

I am not going to mess with GO searches for this.

## Cure/Fail DE plots

It is not surprising that few or no genes are deemed significantly
differentially expressed across samples which were taken from cure or
fail patients.

```{r}
cf_table_nobatch[["plots"]][["fail_vs_cure"]][["deseq_ma_plots"]]

dev <- pp(file = "images/cf_ma.png")
cf_table[["plots"]][["fail_vs_cure"]][["deseq_ma_plots"]]
closed <- dev.off()
cf_table[["plots"]][["fail_vs_cure"]][["deseq_ma_plots"]]
```

# Combining the macrophage infected amastigotes with in-vitro promastigotes

One query we have not yet addressed: what are the similarities and
differences among the strains used to infect the macrophage samples
and the promastigote samples used in the TMRC2 parasite data?

In my container image, this dataset is not currently loaded, so turning this off.

```{r}
## I just fixed this in the datasets Rmd, but until that propagates just set it manually
annotation(lp_expt) <- annotation(lp_macrophage)
tmrc2_macrophage_norm <- normalize_expt(lp_macrophage, transform="log2", convert="cpm",
                                        norm="quant", filter=TRUE)
all_tmrc2 <- combine_expts(lp_expt, lp_macrophage)

all_nosb <- all_tmrc2
pData(all_nosb)[["stage"]] <- "promastigote"
na_idx <- is.na(pData(all_nosb)[["macrophagetreatment"]])
pData(all_nosb)[na_idx, "macrophagetreatment"] <- "undefined"
all_nosb <- subset_expt(all_nosb, subset = "macrophagetreatment!='inf_sb'")
ama_idx <- pData(all_nosb)[["macrophagetreatment"]] == "inf"
pData(all_nosb)[ama_idx, "stage" ] <- "amastigote"
pData(all_nosb)[["batch"]] <- pData(all_nosb)[["stage"]]
```

I think the above picture is sort of the opposite of what we want to
compare in a DE analysis for this set of data, e.g. we want to compare
promastigotes from amastigotes?

```{r}
all_nosb <- set_expt_batches(all_nosb, fact = "condition") %>%
  set_expt_conditions(fact = "stage")
two_zymo <- subset_expt(
  all_nosb,
  subset = "zymodemecategorical=='z22'|zymodemecategorical=='z23'|zymodemecategorical=='unknown'")

pro_ama <- all_pairwise(all_nosb, filter = TRUE, model_batch = "svaseq")
pro_ama_table <- combine_de_tables(
  pro_ama,
  excel = glue("{excel_out}/DE_promastigote_amastigote/{ver}/pro_vs_ama_table-v{ver}.xlsx"))
pro_ama_sig <- extract_significant_genes(
    pro_ama_table,
    excel = glue("{excel_out}/DE_promastigote_amastigote/{ver}/pro_vs_ama_sig-v{ver}.xlsx"))
```

#### Gene ontology comparing the life stages

```{r}
increased_promastigote <- pro_ama_sig[["deseq"]][["ups"]][["promastigote_vs_amastigote"]]
increased_amastigote <- pro_ama_sig[["deseq"]][["downs"]][["promastigote_vs_amastigote"]]
promastigote_goseq <- simple_goseq(increased_promastigote, go_db = lp_go, length_db = lp_lengths)
promastigote_goseq
amastigote_goseq <- simple_goseq(increased_amastigote, go_db = lp_go, length_db = lp_lengths)
amastigote_goseq

## silly, topgo wants the gene id column to be 'ID', I should fix this.
colnames(lp_go) <- c("ID", "GO")
promastigote_topgo <- simple_topgo(increased_promastigote, go_db = lp_go)
enrichplot::dotplot(promastigote_topgo$enrich_results$bp)

amastigote_topgo <- simple_topgo(increased_amastigote, go_db = lp_go)
enrichplot::dotplot(amastigote_topgo$enrich_results$bp)
```

### Plot promastigote/amastigote DE genes

```{r}
pro_ama_table[["plots"]][["promastigote_vs_amastigote"]][["deseq_ma_plots"]]
```

I am a little surprised by this plot, I somewhat expected there to be
few genes which passed the 2-fold difference demarcation line.

```{r}
pander::pander(sessionInfo())
message(paste0("This is hpgltools commit: ", get_git_commit()))
## message(paste0("Saving to ", savefile))
## tmp <- sm(saveme(filename = savefile))
```

```{r, eval=FALSE}
tmp <- loadme(filename = savefile)
```
